File: web/src/app/service/course.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from "@angular/common/http";
import {Observable} from "rxjs";
import {ResponseBody} from "../entity/response-body";
import {schoolIdAndWeeks} from "../entity/schoolIdAndWeeks";
import {SaveRequest} from "../dto/courseDto/saveRequest";
import {GetAllResponse} from "../dto/courseDto/getAllResponse";

@Injectable({
  providedIn: 'root'
})
export class CourseService {
  baseUrl = 'http://localhost:8080/Course';
  constructor(private httpClient: HttpClient) { }

  add(course: SaveRequest): Observable<ResponseBody> {
    return this.httpClient.post<ResponseBody>(`${this.baseUrl}/add`, course);
  }

  delete(courseInfoId: number): Observable<ResponseBody> {
    return this.httpClient.delete<ResponseBody>(`${this.baseUrl}/delete/${courseInfoId}`);
  }

  getAll(): Observable<GetAllResponse[]> {
    return this.httpClient.get<GetAllResponse[]>(`${this.baseUrl}/getAll`);
  }

  getCourseInfoById(id: number): Observable<ResponseBody> {
    return this.httpClient.get<ResponseBody>(`${this.baseUrl}/getCourseInfoById/${id}`);
  }

  // 根据学校和周数，获取该学校所有学生的时间表
  getCourseMessage(params: HttpParams): Observable<ResponseBody> {
    return this.httpClient.get<ResponseBody>(`${this.baseUrl}/getCourseMessage`, {params});
  }

  update(courseInfo: SaveRequest): Observable<ResponseBody> {
    return this.httpClient.patch<ResponseBody>(`${this.baseUrl}/update`, courseInfo);
  }

  // 复用课程安排
  reuse(courseInfoId: number): Observable<ResponseBody> {
    return this.httpClient.post<ResponseBody>(`${this.baseUrl}/reuse/${courseInfoId}`, {});
  }

  // 取消复用
  cancelReuse(courseInfoId: number): Observable<ResponseBody> {
    return this.httpClient.post<ResponseBody>(`${this.baseUrl}/cancelReuse/${courseInfoId}`, {});
  }

  getAllCourseInfo(schoolIdAndWeeksData: schoolIdAndWeeks[]): Observable<ResponseBody> {
    let params = new HttpParams();

    // 将每个 schoolId 和 weeks 添加到查询参数中
    schoolIdAndWeeksData.forEach(item => {
      params = params.append('schoolId', item.schoolId.toString());
      params = params.append('weeks', item.weeks.toString());
    });

    return this.httpClient.get<ResponseBody>(`${this.baseUrl}/getAllCourseInfo`, {params});
  }

  search(searchCourse: string | null, creatorStudent: number | null): Observable<GetAllResponse[]> {
    let params = new HttpParams();

    if (searchCourse) {
      params = params.set('searchCourse', searchCourse);
    }

    if (creatorStudent) {
      params = params.set('creatorStudent', creatorStudent);
    }

    return this.httpClient.get<GetAllResponse[]>(`${this.baseUrl}/search`, { params });
  }
}

File: api/src/main/java/com/mengyunzhi/eduPlanner/service/CourseService.java
package com.mengyunzhi.eduPlanner.service;

import com.mengyunzhi.eduPlanner.dto.CourseDto;
import com.mengyunzhi.eduPlanner.dto.Response;
import com.mengyunzhi.eduPlanner.dto.CurrentUser;
import com.mengyunzhi.eduPlanner.dto.Response;
import com.mengyunzhi.eduPlanner.entity.Course;
import com.mengyunzhi.eduPlanner.entity.CourseInfo;

import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * @author zhangyuxuan
 */
public interface CourseService {
    /**
     * 删除课程安排
     * @param courseInfoId 课程安排id
     * @return 响应
     */
    Response<String> delete(Long courseInfoId);
    /**
     * 新增课程
     * @param saveRequest 新增的课程
     * @param userId 登录用户的id
     * @param schoolId 登录用户的 学校id
     * @return course
     */
    Response<String> save(CourseDto.SaveRequest saveRequest, Long userId, Long schoolId);

    /**
     * 获取课程列表
     * @return 课程安排列表
     */
    List<CourseDto.GetAllCoursesResponse> getAllCourses();

    /**
     * 课程复用
     * @param courseInfoId 被复用的课程安排id
     * @param studentId 需要复用的学生id
     * @return 响应
     */
    Response<String> reuseCourseInfo(Long courseInfoId, Long studentId);

    /**
     * 取消课程复用
     * @param courseInfoId 需要取消的课程安排id
     * @param studentId 操作的学生id
     * @return 响应
     */
    Response<String> cancelReuseCourseInfo(Long courseInfoId, Long studentId);

    /**
     * 检查该课程安排是否还有学生复用
     * @param courseInfoId 该课程id
     * @return response.status = true 未被复用
     */
    Response<String> checkReuseStudent(Long courseInfoId);

    /**
     * 根据school和week获取所有学生的课程安排
     * 如果传了student id，那么获取该周单个学生的课程安排
     * @param schoolId 学校id
     * @param week 周数
     * @param studentId 学生id
     * @return 数组
     */
    Map<Long, List<CourseDto.StudentsCoursesOfSchoolResponse>> getMessage(Long schoolId, Long week, Long studentId);

    /**
     * 获取所有学校所有学生的课程安排信息
     * @param schoolId 学校id列表
     * @param weeks 学校对应的当前周
     * @return 课程安排信息
     */
    Map<Long, List<CourseDto.StudentsCoursesOfSchoolResponse>> getAllMessage(List<Long> schoolId, List<Long> weeks);

    List<CourseDto.GetAllCoursesResponse> search(String searchCourse, Long creatorStudent);

    /**
     * 根据 courseInfoId 获取需要编辑的课程安排
     * @param courseInfoId 待编辑的课程安排
     * @return 响应
     */
    Response<CourseDto.GetCourseInfoByIdResponse> getCourseInfoById(Long courseInfoId);

    /**
     * 更新课程安排
     * @param saveRequest 需要更新的数据
     * @param courseInfoId 更新的课程安排id
     * @param studentId 学生id
     * @return 响应
     */
    Response<String> update(CourseDto.SaveRequest saveRequest, Long courseInfoId, Long studentId);

}